src/app/BufferedReadCallback.cpp

```diff
@@ -0,0 +1,235 @@
+{
+    mCallback.OnReportBegin(apReadClient);
+}
+
+void BufferedReadCallback::OnReportEnd(const ReadClient * apReadClient)
+{
+    CHIP_ERROR err = DispatchBufferedData(nullptr, mBufferedPath, StatusIB(), true);
+    if (err != CHIP_NO_ERROR)
+    {
+        mCallback.OnError(apReadClient, err);
+    }
+
+    mCallback.OnReportEnd(apReadClient);
+}
+
+CHIP_ERROR BufferedReadCallback::GenerateListTLV(System::PacketBufferTLVReader & aReader)
+{
+    System::PacketBufferHandle handle = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSizeWithoutReserve, 0);
+    System::PacketBufferTLVWriter writer;
+    TLV::TLVType outerType;
+
+    //
+    // Initialize the writer, but with chaining enabled, since we'll very likely
+    // exceed the size of a single packet buffer when reconstituing the entire list
+    //
+    writer.Init(std::move(handle), /* useChainedBuffers */ true);
+
+    ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Array, outerType));
+
+    for (auto & bufHandle : mBufferedList)
+    {
+        System::PacketBufferTLVReader reader;
+
+        reader.Init(std::move(bufHandle));
+
+        ReturnErrorOnFailure(reader.Next());
+        ReturnErrorOnFailure(writer.CopyElement(TLV::AnonymousTag, reader));
+    }
+
+    ReturnErrorOnFailure(writer.EndContainer(outerType));
+
+    writer.Finalize(&handle);
+    aReader.Init(std::move(handle));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BufferedReadCallback::BufferData(const ConcreteAttributePath & aPath, TLV::TLVReader * apData)
+{
+
+    if (aPath.mListOp == ConcreteAttributePath::ListOperation::ReplaceAll)
+    {
+        TLV::TLVType outerContainer;
+
+        VerifyOrReturnError(apData->GetType() == TLV::kTLVType_Array, CHIP_ERROR_INVALID_TLV_ELEMENT);
+        mBufferedList.clear();
+
+        ReturnErrorOnFailure(apData->EnterContainer(outerContainer));
+
+        CHIP_ERROR err;
+
+        while ((err = apData->Next()) == CHIP_NO_ERROR)
+        {
+            System::PacketBufferTLVWriter writer;
+            System::PacketBufferHandle handle = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLengthBytes);
+
+            writer.Init(std::move(handle), false);
+
+            ReturnErrorOnFailure(writer.CopyElement(TLV::AnonymousTag, *apData));
+            ReturnErrorOnFailure(writer.Finalize(&handle));
+
+            // Compact the buffer down to a more reasonably sized packet buffer
+            // if we can.
+            //
+            handle.RightSize();
+
+            mBufferedList.push_back(std::move(handle));
+        }
+
+        if (err == CHIP_END_OF_TLV)
+        {
+            err = CHIP_NO_ERROR;
+        }
+
+        ReturnErrorOnFailure(err);
+        ReturnErrorOnFailure(apData->ExitContainer(outerContainer));
+    }
+    else if (aPath.mListOp == ConcreteAttributePath::ListOperation::AppendItem)
+    {
+        System::PacketBufferTLVWriter writer;
```

Comment:
This is a copy/paste of the body of the loop above.  Helper function for copying a single list item over, perhaps?